clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java

```diff
@@ -0,0 +1,260 @@
+        unsentRequests.forEach(this::add);
+    }
+
+    public static class PollResult {
+        public final long timeMsTillNextPoll;
+        public final List<UnsentRequest> unsentRequests;
+
+        public PollResult(final long timeMsTillNextPoll, final List<UnsentRequest> unsentRequests) {
+            this.timeMsTillNextPoll = timeMsTillNextPoll;
+            this.unsentRequests = Collections.unmodifiableList(unsentRequests);
+        }
+    }
+
+    public static class UnsentRequest {
+        private final Optional<AbstractRequestFutureCompletionHandler> callback;
+        private final AbstractRequest.Builder abstractBuilder;
+        private final Optional<Node> node; // empty if random node can be choosen
+        private final Timer timer;
+
+        public UnsentRequest(final Timer timer,
+                             final AbstractRequest.Builder abstractBuilder,
+                             final AbstractRequestFutureCompletionHandler callback) {
+            this(timer, abstractBuilder, callback, null);
+        }
+
+        public UnsentRequest(final Timer timer,
+                             final AbstractRequest.Builder abstractBuilder,
+                             final AbstractRequestFutureCompletionHandler callback,
+                             final Node node) {
+            Objects.requireNonNull(abstractBuilder);
+            this.abstractBuilder = abstractBuilder;
+            this.node = Optional.ofNullable(node);
+            this.callback = Optional.ofNullable(callback);
+            this.timer = timer;
+        }
+
+        public static UnsentRequest makeUnsentRequest(
+                final Timer timeoutTimer,
+                final AbstractRequest.Builder<?> requestBuilder,
+                final AbstractRequestFutureCompletionHandler callback) {
+            return new UnsentRequest(
+                    timeoutTimer,
+                    requestBuilder,
+                    callback);
+        }
+
+        @Override
+        public String toString() {
+            return abstractBuilder.toString();
+        }
+    }
+
+    public static class DefaultRequestFutureCompletionHandler extends AbstractRequestFutureCompletionHandler {
+        @Override
+        public void handleResponse(ClientResponse r, Throwable t) {}
+    }
+
+    public abstract static class AbstractRequestFutureCompletionHandler implements RequestCompletionHandler {
+        private final RequestFuture<ClientResponse> future;
+        private ClientResponse response;
+        private RuntimeException e;
+
+        AbstractRequestFutureCompletionHandler() {
+            this.future = new RequestFuture<>();
+        }
+
+        public void fireCompletion() {
+            if (e != null) {
+                future.raise(e);
+            } else if (response.authenticationException() != null) {
+                future.raise(response.authenticationException());
+            } else if (response.wasDisconnected()) {
```

Comment:
Check if the disconnect was to the current coordinator. If so, then `markCoordinatorUnknown()`.